Fix Discrepancy for publicationYear

sorochak opened the pull request:
This PR fixes the discrepancy outlined in https://github.com/cioos-siooc/metadata-entry-form/issues/295

Updated `recordToDataCite` mapping function to set publicationYear to dateRevised if it exists, otherwise use datePublished.

Updated `generateCitation` function to check for dateRevised before datePublished.

sorochak commented:
I guess this will only fix new records moving forward. Not already submitted data sets.

sorochak commented:
need to fix the citation still.

fostermh reviewed (approved):
this appears to resolve the problem mentioned in the issues but it would be good if tim or nate could also review.

timvdstap commented:
I'm getting an error "Error occurred with DOI API" when I select Update DOI. Also, within the Metadata Entry Tool, the publication year doesn't show up in the recommended citation anymore.

sorochak commented:
@timvdstap I'm unable to reproduce your errors. Let's chat again when you're able to.

timvdstap commented:
We were able to reproduce the error, the revised date showed up correctly when initially generating a draft DOI, only produced an error when trying to update an existing draft DOI.

sorochak commented:
The error occurred because the Date picker returns a Date object when the user selects a new date. 

Firebase stores the date as a string. When clicking the save button, the Date object is updated to a date string.

In `recordToDataCite.js` (our metadata record to Datacite object mapping function), we expect date to be a string, and use .slice() to get the year. The `.slice` method can only be used on a string.  If a user changes the date, and then attempts to update the DOI before saving the record, the `.slice` method then gets called on a Date object instead of a string and throws an error.

Updated `DateInput` to format the date as an ISO string instead of a Date object.
Updated `DateInput.test` to expect ISO string.

timvdstap commented:
publicationYear seems to get updated now properly in DataCite, as well as in the recommended citation in the metadata form. I guess this only applies to newly created metadata records, so I'm not sure yet whether a 'Findable' record in DataCite will also get updated properly in DataCite.

sorochak commented:
should we merge this now then? Or do we need to figure out the `Findable` piece still ?

timvdstap commented:
@sorochak You say that this will only impact new records moving forward, could you tell me why that is? The revision to the publicationYear will also be necessary for time series records that are currently published and are getting a new version. Or do you mean that existing versioned records do not automatically get their publicationYear updated? 

I'm hesitant to try the Findable component through the emulator because a lot of published records in there (such as e.g. the JSP Time Series record) are incomplete compared to their current versions in the Hakai Catalogue. Seems like the only way to test whether publicationYear also gets updated properly for Findable records would be to either a) create a new Findable DOI through the emulator (which we'll have to then set to Registered so it doesn't get indexed), or b) merge this PR and address any bugs that arise.

sorochak commented:
My comment regarding that this change will only impact new records moving forward meant that existing versioned records do not automatically get their publicationYear updated. Those existing records would need to be updated manually for the publicationYear to change to the revision date.
